[ADD] website_sale_product_trial

remytms opened the pull request:
## Description

Module for Trial Subscription.

## Odoo task (if applicable)

[task](https://gestion.coopiteasy.be/web#id=13442&model=project.task&view_type=form&menu_id=536)

## Checklist before approval

- [ ] Tests are present (or not needed).
- [x] Credits/copyright have been changed correctly.
- [ ] Change log snippet is present.
- [x] (If a new module) Moving this to OCA has been considered.

remytms committed:
[ADD] website_sale_product_trial

huguesdk commented on website_sale_product_trial/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1 @@
+Form to order subscription product
```
this doesn’t match the module’s summary.

remytms replied:
changed

huguesdk commented on website_sale_product_trial/i18n/website_sale_restrict_sepa_dd.pot:
this file does not belong in this module.

remytms replied:
removed

huguesdk commented on website_sale_product_trial/__manifest__.py:
```diff
@@ -0,0 +1,27 @@
+# SPDX-FileCopyrightText: 2024 Coop IT Easy SC
```
shouldn’t this be 2025 (and in all other files too)?

remytms replied:
I think I created this module in 2024, but I will change to 2025.

huguesdk commented on website_sale_product_trial/models/sale_order.py:
```diff
@@ -0,0 +1,48 @@
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.constrains("order_line")
+    def _check_trial_alone(self):
+        """Ensure trial are not mixed in a sale order."""
+        for order in self:
+            at_least_one_trial = any(
+                order.order_line.mapped("product_id").mapped("is_trial")
+            )
+            if at_least_one_trial and order.order_line > 1:
+                raise ValidationError(
+                    _(
+                        "Cannot add product trial in an order that "
+                        "contains other products."
+                    )
+                )
+
+    def check_product_compatibility(self, product_id):
+        warning = super().check_product_compatibility(product_id)
+        if not warning:
+            product = self.env["product.product"].browse(product_id).exists()
+            non_trials = self.order_line.mapped("product_id").filtered(
+                lambda p: not p.is_trial
+            )
+            any_trial = any(self.order_line.mapped("product_id").mapped("is_trial"))
+            if product:
+                if product.is_trial and non_trials:
+                    warning = _(
+                        f"Product {product.name} cannot be added because "
+                        "it's a trial and trial must be ordered seperately."
+                    )
+                elif not product.is_trial and any_trial:
+                    warning = _(
+                        f"Product {product.name} cannot be added because "
+                        "order contains products that are trials and "
+                        "trials must be ordered separately."
+                    )
+        return warning
```
same remarks as in the `website_sale_product_contract_gift` ([in #351](https://github.com/coopiteasy/addons/pull/351/files/01126faf58b4ae3515c785688a07dfe67eec77a4#r1985131776)).

remytms replied:
It’s done.

remytms committed:
[ADD] website_sale_product_trial: 14 day duration for trial products

Hard coding this value is not a good idea. But is a trade of cost and
flexibility.

huguesdk commented on website_sale_product_trial/models/sale_order.py:
```diff
@@ -46,3 +47,25 @@ def check_product_compatibility(self, product_id):
+        product = self.env["product.product"].browse(product_id)
+        if product.is_trial:
+            values["date_start"] = datetime.today()
+            values["date_end"] = datetime.today() + timedelta(weeks=2)
```
please use a constant defined above, like:

```python
TRIAL_DURATION_DAYS = 14
```

remytms replied:
done

remytms replied:
done

remytms committed:
[IMP] website_sale_product_trial: copyright and description

remytms committed:
[FIX] website_sale_product_trial: fix compatibility with delivery

remytms commented:
I added test to verify product compatibility checks works.

huguesdk committed:
[FIX] fix english string

huguesdk reviewed (approved):
all lgtm now! thanks, @remytms!

huguesdk commented:
/ocabot merge nobump
